src/extra/components/pulses.py

```diff
@@ -940,15 +959,115 @@ def _identify_ppl_seed(cls, data):
     def _mask_table(self, table):
         return is_laser(table, self._ppl_seed)
 
-    def _get_ppdecoder_node(self):
-        return 'laser'
+    def _get_ppdecoder_nodes(self):
+        return ['laser']
 
     @property
     def ppl_seed(self) -> Optional[PPL_BITS]:
         """Used laser seed."""
         return self._ppl_seed
 
 
+class MachinePulses(TimeserverPulses):
+    """An interface to machine pulses and other bunch pattern data.
+
+    The bunch pattern table contains much more information than just the
+    pulses of each SASE beamline or the optical laser systems, in
+    particular about the upstream electron systems.
+
+    This component is configurable to extract any information from the
+    bunch pattern table beyond the X-ray or optical laser pulses. By
+    default, it considers all generated electron bunches by combining
+    the various electron dumps along the accelerator. Please consult the
+    XFEL Timing System Specification for the various mask bits used.
+
+    Note that the full range of bunch pattern table information is only
+    available when used with timeserver data. The pulse pattern decoder
+    data only contains information about those pulses in the main dump
+    and SASE dumps.
+
+    Args:
+        data (extra.data.DataCollection): Data to access bunch pattern
+            data from.
+        source (str, optional): Source name of a timeserver or pulse
+            pattern decoder, only needed if the data includes more than
+            one such device or none could not be detected automatically.
+        mask (int, optional): Custom mask into the bunch pattern table,
+            by default all electron dumps are used. When initialized
+            with ppdecoder data, only the main and SASE dumps are
+            supported.
+        require_all_bits (bool, optional): Whether to consider entries
+            having any or all of the bits specified by mask, False by
+            default. Only supported when initialized with timeserver.
```

Comment:
Nitpicking:
```suggestion
            default. Only supported when initialized with a timeserver.
```

Reply:
Ah, that was supposed to be `with timeserver data.`